### DISCUSSION TITLE
Enhancement Proposal: Adding Intent Field to Context Object

### DISCUSSION THREAD

**Discussion by klatka** [home-assistant/architecture#1067] [General]
tl;dr Add another field (optional) to the context object:

| Field  | Description                                                                                               |
| ------ | ----------------------------------------------------------------------------- |
| intent | Custom identifier passed through from the corresponding service call. |

## Summary
This proposal suggests adding an optional "intent" field to the context object in order to provide additional context to events triggered by service calls. By including this field, services and workflows can pass custom identifiers that help in understanding the purpose or source of the event, thereby facilitating more efficient event handling.

## Problem Statement
In scenarios where events are triggered by service calls without direct user interaction, extracting meaningful context from the event can be challenging. This becomes particularly evident in automated systems like lighting control, where distinguishing between manual and automated changes is crucial for proper event handling.

## Proposed Solution
The proposal recommends introducing an "intent" field to the context object associated with service calls. This field can hold custom identifiers passed from the service call, providing insights into the purpose or source of the event. Including this field allows for more informed event processing, especially in automated systems where discerning user intent is essential.

## Example Use Cases
1. **Automated Lighting System:** 
   - When the lighting system receives a "state_changed" event, it checks for the presence of the "intent: auto_lighting" attribute.
   - If the event lacks this attribute, it identifies the change as manual and adjusts its behavior accordingly.
   - This enables the system to differentiate between manual and automated light adjustments, ensuring proper handling in each scenario.

2. **Presence Simulation:**
   - In scenarios where presence simulation is employed, the intent field helps in distinguishing between simulated and manual light activations.
   - By recognizing the intent behind the service call, the lighting system can adjust its response accordingly, maintaining accurate simulation.

## Benefits

1. **Improved Event Handling:** 
   - The addition of the "intent" field enhances the system's ability to understand the purpose or source of events triggered by service calls.
   - This leads to more accurate event processing and response, reducing the likelihood of erroneous actions.

2. **Enhanced Automation:** 
   - Automated systems can leverage the intent field to differentiate between user-initiated and system-generated changes, enabling more sophisticated automation logic.

3. **Reduced Processing Overhead:**
   - By providing clear intent information upfront, the proposal aims to streamline event processing, potentially reducing CPU overhead, especially in complex systems.

## Example

```
service: light.turn_on
target:
  entity_id: light.hue_color_lamp_1
intent: auto_lighting
```

And the event fired would look like this:

```
event_type: state_changed
data: {...}
origin: LOCAL
time_fired: "2024-03-26T10:43:06.920615+00:00"
context:
  id: 01HN2N94K8JNG9JYGRVTPQWT1B
  parent_id: null
  user_id: null
  intent: auto_lighting
```

## Conclusion
Introducing the "intent" field to the context object offers a straightforward yet powerful enhancement to event processing in automated systems. By providing valuable context information upfront, this proposal aims to improve efficiency, accuracy, and flexibility in handling events triggered by service calls.

**Comment by balloob** [comment]
Home Assistant allows this today by firing plain events. For example, an automation fires `automation_fired` with relevant information such that all the context can be traced back that the changes came from the automation.

**Reply by klatka** [reply]
I would reject tracing service calls by firing plain events as a solution because of the following downsides compared to the approach of this proposal.

The automated lighting system (could exist of n automations or flows for example)

1. has to make sure to fire the plain event first before doing service calls.
2. needs to cache context ids globally before making service calls to know which are from this automated system (a lifecycle for this cache needs to be implemented too).
3. which is also listening to `state_changed` events has to check all cached context ids first to know if the event is fired as a consequence of a service call from itself to handle the event properly.
4. can run into race condition issues (imagine a home with a lot of people where it's possible to have two or more motion-activated areas at the same time).
5. processing overhead compared to just adding an intent to all service calls (the goal of this proposal).

**Reply by balloob** [reply]
What I am saying is that we already have this. Look at the logbook when an automation fires and how it is already attributing things correctly:

![CleanShot 2024-03-29 at 10 53 12@2x](https://github.com/home-assistant/architecture/assets/1444314/850b1f03-e5a5-4409-8502-b70be2af1ad6)

**Reply by klatka** [reply]
Is this only working in the automation scope or is it possible to use this tracing behavior to service calls from other sources like components or node-red? For example, light has changed color triggered by component/custom_component x

**Reply by evelant** [reply]
IIRC there's another problem with context. I think state changes get attributed to a service call via a time window of 5 seconds whether or not that service call actually initiated the state change. It's just assumed that any state changes within 5s are due to whatever the most recent service call was (idk if this got improved but I think that's how it was last time I looked). This means that it's more or less impossible to know what actually triggered a state change because it's easily misattributed if there are multiple service calls. I think https://github.com/home-assistant/architecture/discussions/1059 also deals with this.

**Reply by klatka** [reply]
If someone is running through the house and 4 motion sensors are triggered which results in 4 `light.turn_on` service calls while some other user decides to switch lights in his room, all within 5 seconds, how is it possible to map each state change to the right service call?

In addition the user could use an ESPHome switch for example that makes the service call, so there is no user_id in context.

**Reply by evelant** [reply]
It isn't possible right now AFAIK, I think you'll get wrong attribution data in that scenario. I think it would be a very useful enhancement to have precise (and overall more) context data available.

**Reply by klatka** [reply]
I agree. We need more context data :) The idea of sending events to trace service calls makes things way too complicated and doesn't work in all contexts.

**Reply by klatka** [reply]
@balloob Would you accept a PR that adds an optional property to extend the possibilities providing context?
